File: backend/src/canonical/amp_detect.rs
//! AMP URL detection.
//!
//! Substring scan over **parsed URL components** (host / path / query) so a
//! whole-URL match like `https://amputeestore.com/...` doesn't false-positive
//! on `/amp` just because of the `//amp` after the scheme.

use url::Url;

/// Scanned against each URL component by [`is_amp_url`].
pub(crate) const AMP_KEYWORDS: &[&str] = &[
    "/amp", "amp/", ".amp", "amp.", "?amp", "amp?", "=amp", "amp=", "&amp", "amp&", "%amp", "amp%",
    "_amp", "amp_",
];

/// Domains that the keyword scan misfires on (band+amp, etc.).
const DENYLISTED_DOMAINS: &[&str] = &[
    "video.twimg.kim",
    "bandcamp.com",
    "progonlymusic.com",
    "redd.it",
    "reddit.com",
    "spotify.com",
    "youtube.com",
    "youtu.be",
];

/// `true` if any of [`AMP_KEYWORDS`] appears as a substring in the URL's
/// host, path, or query — unless the host is on [`DENYLISTED_DOMAINS`].
pub fn is_amp_url(url: &str) -> bool {
    let Ok(parsed) = Url::parse(url) else {
        return false;
    };

    let host = parsed.host_str().unwrap_or("").to_ascii_lowercase();
    if DENYLISTED_DOMAINS
        .iter()
        .any(|d| host_matches_domain(&host, d))
    {
        return false;
    }

    if has_amp_keyword(&host) {
        return true;
    }

    let path = parsed.path().to_ascii_lowercase();
    if has_amp_keyword(&path) {
        return true;
    }

    if let Some(query) = parsed.query() {
        let query = query.to_ascii_lowercase();
        if has_amp_keyword(&query) {
            return true;
        }
    }

    false
}

fn has_amp_keyword(s: &str) -> bool {
    AMP_KEYWORDS.iter().any(|kw| s.contains(kw))
}

/// `true` when `host` is exactly `domain` or a real subdomain of it.
///
/// Plain `host.ends_with(domain)` would also match unrelated hosts like
/// `notyoutube.com` ↔ `youtube.com` or `notampproject.org` ↔ `ampproject.org`,
/// so denylist / cache-host checks need a dot-boundary on the suffix side.
fn host_matches_domain(host: &str, domain: &str) -> bool {
    host == domain
        || (host.len() > domain.len()
            && host.ends_with(domain)
            && host.as_bytes()[host.len() - domain.len() - 1] == b'.')
}

/// Returns `true` if the URL is hosted on a known AMP cache
/// (Google AMP, Bing AMP, or `ampproject.{net,org}`).
///
/// Ports `praw-python-archive/helpers/checker_utils.py:check_if_cached`. Like
/// [`is_amp_url`], we evaluate against parsed components rather than the
/// raw string — gives the same answers but is harder to fool.
pub fn is_cached_amp(url: &str) -> bool {
    let Ok(parsed) = Url::parse(url) else {
        return false;
    };

    let host = parsed.host_str().unwrap_or("").to_ascii_lowercase();
    let path = parsed.path().to_ascii_lowercase();

    if host_matches_domain(&host, "ampproject.net") || host_matches_domain(&host, "ampproject.org")
    {
        return true;
    }

    let on_google = host.starts_with("www.google.");
    let on_bing = host.starts_with("www.bing.");
    if (on_google || on_bing) && path.starts_with("/amp/") {
        return true;
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detects_google_amp_cache_url() {
        assert!(is_amp_url(
            "https://www.google.com/amp/s/electrek.co/2018/06/19/tesla-model-3/amp/"
        ));
    }

    #[test]
    fn detects_publisher_amp_path() {
        assert!(is_amp_url(
            "https://www.bbc.com/news/world-europe-12345/amp"
        ));
        assert!(is_amp_url("https://www.fox5atlanta.com/video/foo.amp"));
    }

    #[test]
    fn detects_amp_subdomain() {
        assert!(is_amp_url("https://amp.cnn.com/cnn/2020/some-article"));
    }

    #[test]
    fn detects_amp_query_param() {
        assert!(is_amp_url("https://example.eu/article?amp=1"));
        assert!(is_amp_url("https://example.eu/article?output=amp"));
    }

    #[test]
    fn detects_ampproject_subdomain() {
        assert!(is_amp_url(
            "https://www-cnn-com.cdn.ampproject.org/c/s/www.cnn.com/sample"
        ));
    }

    #[test]
    fn rejects_amputeestore_false_positive() {
        // Canonical false-positive case: a whole-URL substring scan flags
        // `/amp` inside `//amputeestore.com` after the scheme.
        assert!(!is_amp_url(
            "https://amputeestore.com/collections/prosthetic-socks/products/knit-rite-liner-liner-sock?variant=4114742017"
        ));
        assert!(!is_amp_url(
            "https://amputeestore.com/products/tamarack-glidewear-prosthetic-liner-patch"
        ));
        assert!(!is_amp_url(
            "https://amputeestore.com/products/alps-antiperspirant-spray"
        ));
    }

    #[test]
    fn rejects_non_amp_urls() {
        assert!(!is_amp_url("https://www.google.com/search?q=foo"));
        assert!(!is_amp_url("https://news.ycombinator.com/item?id=42"));
        assert!(!is_amp_url("https://en.wikipedia.org/wiki/Wikipedia"));
    }

    #[test]
    fn rejects_denylisted_domains_even_with_amp_in_path() {
        assert!(!is_amp_url("https://www.youtube.com/amp/some-video"));
        assert!(!is_amp_url("https://open.spotify.com/amp/track/123"));
        assert!(!is_amp_url("https://bandcamp.com/amp/foo"));
    }

    #[test]
    fn denylist_does_not_match_unrelated_hosts() {
        // `notyoutube.com` ends with `youtube.com` but isn't a subdomain.
        assert!(is_amp_url("https://notyoutube.com/amp/some-video"));
        assert!(is_amp_url("https://myreddit.com/article/amp"));
    }

    #[test]
    fn cached_does_not_match_unrelated_hosts() {
        // `notampproject.org` ends with `ampproject.org` but is not a real
        // subdomain — must NOT be classified as an AMP cache.
        assert!(!is_cached_amp("https://notampproject.org/some/path"));
        assert!(!is_cached_amp("https://fakeampproject.net/x"));
    }

    #[test]
    fn rejects_malformed_urls() {
        assert!(!is_amp_url("not a url"));
        assert!(!is_amp_url(""));
        assert!(!is_amp_url("amp"));
    }

    #[test]
    fn cached_detects_google_amp() {
        assert!(is_cached_amp(
            "https://www.google.com/amp/s/electrek.co/2018/06/19/tesla-model-3/amp/"
        ));
        assert!(is_cached_amp(
            "https://www.google.co.uk/amp/s/example.eu/article"
        ));
    }

    #[test]
    fn cached_detects_bing_amp() {
        assert!(is_cached_amp(
            "https://www.bing.com/amp/s/example.eu/article"
        ));
    }

    #[test]
    fn cached_detects_ampproject() {
        assert!(is_cached_amp("https://cdn.ampproject.org/c/s/example.eu"));
        assert!(is_cached_amp(
            "https://www-cnn-com.cdn.ampproject.org/c/s/foo"
        ));
        assert!(is_cached_amp("https://example.ampproject.net/some/path"));
    }

    #[test]
    fn cached_rejects_publisher_amp_pages() {
        // A publisher's own /amp page is AMP but not "cached" on a third-party CDN.
        assert!(!is_cached_amp("https://www.bbc.com/news/world-europe/amp"));
        assert!(!is_cached_amp("https://amp.cnn.com/cnn/2020/article"));
    }
}

File: devvit-app/src/server/core/ampDetect.ts
// AMP URL detection.
//
// Mirrors `backend/src/canonical/amp_detect.rs` — same 14 keywords, same denylist,
// same component-scoped scan (host / path / query checked separately). The Rust
// impl is the source of truth; if the two ever drift the resolver wins and this
// file gets updated to match.
//
// Why component-scoped: the legacy Python scanned the whole URL string with 14
// short keywords, which false-positived on `amputeestore.com` etc. (`/amp`
// matched `//amp` after the scheme). Parsing first and matching per-component
// fixes that without losing legitimate matches.

// 14 substring patterns from `archive/static/static.py:8-9`.
// Exported because GUESS_AND_CHECK mutates URLs by removing each keyword in
// turn — that code is in the Rust backend, but keeping the export here makes
// the parallel visible.
export const AMP_KEYWORDS = [
  '/amp',
  'amp/',
  '.amp',
  'amp.',
  '?amp',
  'amp?',
  '=amp',
  'amp=',
  '&amp',
  'amp&',
  '%amp',
  'amp%',
  '_amp',
  'amp_',
] as const;

// Domains hard-excluded from AMP detection regardless of URL shape.
// Ports `archive/static/static.py:10`.
const DENYLISTED_DOMAINS = [
  'video.twimg.kim',
  'bandcamp.com',
  'progonlymusic.com',
  'redd.it',
  'reddit.com',
  'spotify.com',
  'youtube.com',
  'youtu.be',
] as const;

function tryParseUrl(input: string): URL | null {
  try {
    return new URL(input);
  } catch {
    return null;
  }
}

function hasAmpKeyword(s: string): boolean {
  return AMP_KEYWORDS.some((kw) => s.includes(kw));
}

// Returns true if the URL appears to be an AMP URL.
//
// Rules (same as Rust):
//   1. Must parse as a URL.
//   2. Host gets checked against DENYLISTED_DOMAINS first — denied → not AMP.
//   3. Host, path, and query are each scanned for AMP_KEYWORDS. Any one hit → AMP.
export function isAmpUrl(input: string): boolean {
  const parsed = tryParseUrl(input);
  if (!parsed) return false;

  const host = parsed.hostname.toLowerCase();
  if (DENYLISTED_DOMAINS.some((d) => host.endsWith(d))) return false;

  if (hasAmpKeyword(host)) return true;

  const path = parsed.pathname.toLowerCase();
  if (hasAmpKeyword(path)) return true;

  // `search` includes the leading `?`; strip it so `?amp` and `amp?` keywords
  // don't both match the same `?amp=1` query (Rust uses `url.query()` which
  // returns the bare query string for the same reason).
  if (parsed.search.length > 1) {
    const query = parsed.search.slice(1).toLowerCase();
    if (hasAmpKeyword(query)) return true;
  }

  return false;
}

// Returns true if the URL is hosted on a known AMP cache
// (Google AMP, Bing AMP, or ampproject.{net,org}).
// Ports `archive/helpers/checker_utils.py:check_if_cached`.
export function isCachedAmp(input: string): boolean {
  const parsed = tryParseUrl(input);
  if (!parsed) return false;

  const host = parsed.hostname.toLowerCase();
  const path = parsed.pathname.toLowerCase();

  if (host.endsWith('ampproject.net') || host.endsWith('ampproject.org')) {
    return true;
  }

  const onGoogle = host.startsWith('www.google.');
  const onBing = host.startsWith('www.bing.');
  if ((onGoogle || onBing) && path.startsWith('/amp/')) {
    return true;
  }

  return false;
}
